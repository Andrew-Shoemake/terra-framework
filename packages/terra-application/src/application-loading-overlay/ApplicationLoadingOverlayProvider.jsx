import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import Scroll from 'terra-scroll';
import LoadingOverlay from 'terra-overlay/lib/LoadingOverlay';
import OverlayContainer from 'terra-overlay/lib/OverlayContainer';
import ApplicationLoadingOverlayContext from './ApplicationLoadingOverlayContext';

const propTypes = {
  /**
   * The children to be rendered within the context of the ApplicationLoadingOverlayProvider.
   * Components rendered here are able to interact with ApplicationLoadingOverlayProvider through
   * the ApplicationLoadingOverlayContext.
   */
  children: PropTypes.node,
  /**
   * A function that is expected to return a custom instance of a loading overlay. If not provided, the
   * default LoadingOverlay component will be rendered directly.
   * Ex. renderLoadingOverlay(Object registeredOverlays)
   */
  renderLoadingOverlay: PropTypes.func,
  /**
   * A function to be called with the current ref of the scrollable element rendered within the
   * ApplicationLoadingOverlayProvider.
   */
  scrollRefCallback: PropTypes.func,
};

const defaultProps = {
  children: undefined,
  scrollRefCallback: undefined,
};

const ApplicationLoadingOverlayProvider = ({
  children,
  renderLoadingOverlay,
  scrollRefCallback,
  ...customProps
}) => {
  const [loadingIndicators, setLoadingIndicators] = useState({});

  const contextValue = useMemo(() => ({
    showLoadingIndicator: (key, message) => {
      setLoadingIndicators(state => (
        {
          ...state,
          [`${key}`]: message,
        }
      ));
    },
    hideLoadingIndicator: (key) => {
      setLoadingIndicators((state) => {
        const newLoadingIndicators = { ...state };
        delete newLoadingIndicators[key];

        return newLoadingIndicators;
      });
    },
  }), []);

  let overlayContent;
  if (renderLoadingOverlay) {
    overlayContent = renderLoadingOverlay(loadingIndicators);
  } else {
    overlayContent = (
      <LoadingOverlay
        isRelativeToContainer
        isAnimated
        isOpen={!!Object.keys(loadingIndicators).length}
        message={Object.keys(loadingIndicators).map(key => (
          loadingIndicators[key]
        )).join(', ')}
      />
    );
  }

  return (
    <OverlayContainer
      {...customProps}
      style={{ height: '100%' }}
      overlay={overlayContent}
    >
      <Scroll refCallback={scrollRefCallback}>
        <ApplicationLoadingOverlayContext.Provider value={contextValue}>
          {children}
        </ApplicationLoadingOverlayContext.Provider>
      </Scroll>
    </OverlayContainer>
  );
};

ApplicationLoadingOverlayProvider.propTypes = propTypes;
ApplicationLoadingOverlayProvider.defaultProps = defaultProps;

export default ApplicationLoadingOverlayProvider;
