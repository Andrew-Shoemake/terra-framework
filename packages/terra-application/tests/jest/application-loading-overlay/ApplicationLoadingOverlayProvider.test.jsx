import React from 'react';
import ApplicationLoadingOverlayProvider from '../../../src/application-loading-overlay/ApplicationLoadingOverlayProvider';

describe('ApplicationLoadingOverlayProvider', () => {
  describe('Snapshots', () => {
    it('should render with minimal props', () => {
      const wrapper = shallow((
        <ApplicationLoadingOverlayProvider />
      ));

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with optional props', () => {
      const scrollRefCallback = jest.fn();
      const wrapper = shallow((
        <ApplicationLoadingOverlayProvider
          scrollRefCallback={scrollRefCallback}
        >
          <div>Test child</div>
        </ApplicationLoadingOverlayProvider>
      ));

      expect(wrapper).toMatchSnapshot();
    });

    it('should render the loading overlay based on current state', () => {
      const reactUseState = React.useState;
      const mockSetState = jest.fn();
      React.useState = () => ([{
        1: 'Message 1',
        2: 'Message 2',
      }, mockSetState]);

      const wrapper = shallow((
        <ApplicationLoadingOverlayProvider>
          <div>Test child</div>
        </ApplicationLoadingOverlayProvider>
      ));

      expect(wrapper).toMatchSnapshot();

      React.useState = reactUseState;
    });
  });
});
