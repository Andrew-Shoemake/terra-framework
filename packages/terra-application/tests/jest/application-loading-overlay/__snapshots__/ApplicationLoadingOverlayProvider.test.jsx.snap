// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationLoadingOverlayProvider Snapshots should render the loading overlay based on current state 1`] = `
<OverlayContainer
  className="container"
  overlay={
    <LoadingOverlay
      backgroundStyle="light"
      isAnimated={true}
      isOpen={true}
      isRelativeToContainer={true}
      message="Message 1, Message 2"
      rootSelector="#root"
    />
  }
>
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</OverlayContainer>
`;

exports[`ApplicationLoadingOverlayProvider Snapshots should render with minimal props 1`] = `
<OverlayContainer
  className="container"
  overlay={
    <LoadingOverlay
      backgroundStyle="light"
      isAnimated={true}
      isOpen={false}
      isRelativeToContainer={true}
      message=""
      rootSelector="#root"
    />
  }
>
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    />
  </Scroll>
</OverlayContainer>
`;

exports[`ApplicationLoadingOverlayProvider Snapshots should render with optional props 1`] = `
<OverlayContainer
  className="container"
  overlay={
    <LoadingOverlay
      backgroundStyle="light"
      isAnimated={true}
      isOpen={false}
      isRelativeToContainer={true}
      message=""
      rootSelector="#root"
    />
  }
>
  <Scroll
    refCallback={[MockFunction]}
  >
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</OverlayContainer>
`;
